src/main/java/com/aws/greengrass/shadowmanager/PubSubIntegrator.java

```diff
@@ -0,0 +1,173 @@
+    private final GetThingShadowRequestHandler getThingShadowRequestHandler;
+    private final PubSubClientWrapper pubSubClientWrapper;
+    private final Pattern shadowPattern = Pattern.compile("\\$aws\\/things\\/(.*)\\/shadow(\\/name\\/(.*))?"
+            + "\\/(update|delete|get)");
+    private final AtomicBoolean subscribed = new AtomicBoolean(false);
+
+    /**
+     * Constructor for PubSubIntegrator.
+     *
+     * @param pubSubClientWrapper             The PubSub client wrapper
+     * @param deleteThingShadowRequestHandler handler class to handle the Delete Shadow request.
+     * @param updateThingShadowRequestHandler handler class to handle the Update Shadow request.
+     * @param getThingShadowRequestHandler    handler class to handle the Get Shadow request.
+     */
+    public PubSubIntegrator(PubSubClientWrapper pubSubClientWrapper,
+                            DeleteThingShadowRequestHandler deleteThingShadowRequestHandler,
+                            UpdateThingShadowRequestHandler updateThingShadowRequestHandler,
+                            GetThingShadowRequestHandler getThingShadowRequestHandler) {
+        this.pubSubClientWrapper = pubSubClientWrapper;
+        this.deleteThingShadowRequestHandler = deleteThingShadowRequestHandler;
+        this.updateThingShadowRequestHandler = updateThingShadowRequestHandler;
+        this.getThingShadowRequestHandler = getThingShadowRequestHandler;
+    }
+
+    /**
+     * Subscribes to the shadow topic over local PubSub.
+     */
+    public void subscribe() {
+        if (this.subscribed.compareAndSet(false, true)) {
+            this.pubSubClientWrapper.subscribe(this::handlePublishedMessage);
+        }
+    }
+
+    /**
+     * Unsubscribes to the shadow topic over local PubSub.
+     */
+    public void unsubscribe() {
+        if (this.subscribed.compareAndSet(true, false)) {
+            this.pubSubClientWrapper.unsubscribe(this::handlePublishedMessage);
+        }
+
+    }
+
+    /**
+     * Helper function to extract the thingName, shadowName and shadow operation from mqtt topic.
+     *
+     * @param topic PubSub topic on which the message was sent.
+     * @return ShadowRequest object with shadow details
+     */
+    ShadowRequest extractShadowFromTopic(String topic) {
+        final Matcher matcher = shadowPattern.matcher(topic);
+
+        if (matcher.find() && matcher.groupCount() == 4) {
+            String thingName = matcher.group(1);
+            String shadowName = matcher.group(3);
+            String operation = matcher.group(4);
+            return new ShadowRequest(thingName, shadowName, operation);
+        }
+        logger.atWarn()
+                .kv("topic", topic)
+                .log("Unable to parse shadow topic for thing name, shadow name and shadow operation");
+        throw new IllegalArgumentException("Unable to parse shadow topic for thing name shadow name "
+                + "and shadow operation");
```

Comment:
nit: "Unable to parse shadow topic" should be fine - otherwise add a comma between 'thing name, shadow name, and shadow operation'

Reply:
+1 for "Unable to parse shadow topic". Operation is optional and may lead to confusion if user did not mean to supply an operation.